File: scripts/preprocessing/gee/javascript/landsat_time_series.js
/*
 * ---
 * title: "Landsat Time Series Analysis"
 * author: "Brendan Casey"
 * created: "2024-12-05"
 * description: Generates a time series of Landsat satellite imagery,
 * calculates user-defined spectral indices, and outputs results as
 * multiband images for further analysis. 
 * ---
 */

/* 1. Setup
 * Prepare the environment, including the AOI, helper functions,
 * and date list for time series processing.
 */

/* Load helper functions */
var utils = require(
  "users/bgcasey/science_centre:functions/utils"
  );
var landsatTimeSeries = require(
  "users/bgcasey/science_centre:functions/landsat_time_series"
  );
var landsatIndicesAndMasks = require(
  "users/bgcasey/science_centre:functions/landsat_indices_and_masks"
  );

/* Define area of interest (AOI) */
var aoi = ee.FeatureCollection('FAO/GAUL_SIMPLIFIED_500m/2015/level1')
  .filter(ee.Filter.eq('ADM0_NAME', 'Canada'))
  .filter(ee.Filter.eq('ADM1_NAME', 'Alberta'))
  .geometry()

/* Small aoi for testing purposes */
// var aoi = ee.Geometry.Polygon([
//   [-113.5, 55.5],  // Top-left corner
//   [-113.5, 55.0],  // Bottom-left corner
//   [-112.8, 55.0],  // Bottom-right corner
//   [-112.8, 55.5]   // Top-right corner
// ]);

/* Create a date list 
 * The date list specifies the starting points for time 
 * intervals used to extract a time series. The createDateList 
 * function generates a list of dates at a specified interval 
 * (e.g., 1 year), beginning on the provided start date 
 * ('2000-06-01') and ending on the end date ('2024-06-01').
 *
 * For each date in the list, the ls_fn function will create a 
 * new end date by advancing the start date by a user-defined 
 * number of time units (e.g., 4 months, 6 weeks). Indices will 
 * be calculated for each of these time intervals.
 *
 * Due to memory limits when generating a time series of
 * Alberta wide images, time series are generated in five year
 * batches. Comment out the unused time periods.  
 */

// var dateList = utils.createDateList(
//   ee.Date('2001-06-01'), ee.Date('2005-06-01'), 1, 'years'
// );

// var dateList = utils.createDateList(
//   ee.Date('2006-06-01'), ee.Date('2010-06-01'), 1, 'years'
// );

// var dateList = utils.createDateList(
//   ee.Date('2011-06-01'), ee.Date('2015-06-01'), 1, 'years'
// );

// var dateList = utils.createDateList(
//   ee.Date('2016-06-01'), ee.Date('2020-06-01'), 1, 'years'
// );

var dateList = utils.createDateList(
  ee.Date('2000-06-01'), ee.Date('2024-06-01'), 1, 'years'
);
 
print("Start Dates", dateList);

/* Define reducer statistic */
var statistic = 'mean'; // Choose from 'mean', 'median', 'max', etc.


/* 2. Landsat Time Series Processing
 * Calculate user-defined spectral indices for Landsat imagery.
 * 
 * Available indices:
 * - BSI: Bare Soil Index
 * - DRS: Distance Red & SWIR
 * - DSWI: Disease Stress Water Index
 * - EVI: Enhanced Vegetation Index
 * - GNDVI: Green Normalized Difference Vegetation Index
 * - LAI: Leaf Area Index
 * - NBR: Normalized Burn Ratio
 * - NDMI: Normalized Difference Moisture Index
 * - NDSI: Normalized Difference Snow Index
 * - NDVI: Normalized Difference Vegetation Index
 * - NDWI: Normalized Difference Water Index
 * - SAVI: Soil Adjusted Vegetation Index
 * - SI: Shadow Index
 */
var ls = landsatTimeSeries.ls_fn(
  dateList, 121, 'days', aoi,
  [
    'BSI', 'DRS', 'DSWI', 'EVI', 'GNDVI', 
    'LAI', 'NBR', 'NDMI', 'NDSI', 'NDVI',
    'NDWI', 'SAVI', 'SI'
  ],
  statistic
)
  // // Apply NDRS for Conifer
  // .map(function(image) {
  //   return landsatIndicesAndMasks.addNDRS(image, [210]);
  // })
  // // Apply NDRS for Broadleaf
  // .map(function(image) {
  //   return landsatIndicesAndMasks.addNDRS(image, [220]);
  // })
  // // Apply NDRS for all forest types
  // .map(function(image) {
  //   return landsatIndicesAndMasks.addNDRS(image);
  // })
  .map(function(image) {
    // Exclude QA_PIXEL band and rename remaining bands
    var filteredBandNames = image.bandNames().filter(
      ee.Filter.neq('item', 'QA_PIXEL')
    );
    return image
      .select(filteredBandNames)
      .toFloat(); // Convert all bands to Float32
  });

//print("Landsat Time Series:", ls);

/* 3. Check Calculated Bands
 * Review to make sure calculations and indices appear 
 * correct.
 */

/* 3.1 Check band summary statistics
* For each band calculate the min, max, and 
* standard deviation of pixel values and print to console.
* Check for values outside the expected range.
*/

/* Calculate summary statistics for the image collection */
var reducer = ee.Reducer.min()
  .combine(ee.Reducer.max(), '', true);

var collectionStats = utils.calculateImageCollectionStats(ls, 
                                                        aoi, 
                                                        1000, 
                                                        1e13, 
                                                        reducer);
print(collectionStats, "image stats")
utils.exportStatsToCSV(collectionStats, 'image_stats');



// /* 3.2 Plot NDVI
// * Visualize the NDVI for 2024 by adding it to the map.
// * Set visualization parameters to highlight vegetation health.
// */

// Define visualization parameters for NDVI
var ndviVisParams = {
  min: -0.5, // Lower limit for NDVI values
  max: 1.0,  // Upper limit for NDVI values
  palette: ['blue', 'white', 'green'] // Color palette
};

var image_first = ls.first();

// Extract the NDVI band from the 2024 image
var ndvi_first = image_first.select('NDVI');

// Reduce resolution for visualization
var ndvi_firstLowRes = ndvi_first.reproject({
  crs: ndvi_first.projection(),
  scale: 2000
});

// Center the map on the area of interest (AOI)
Map.centerObject(aoi, 9);

// Add the low-resolution NDVI layer to the map
Map.addLayer(ndvi_first, ndviVisParams, 'NDVI (Low Res)');


// // Create a mask where valid (non-missing) pixels are 1, and missing pixels are 0
// var missingPixelsMask = ndvi_first.mask().not();

// // Visualize the missing pixels
// Map.addLayer(missingPixelsMask, {min: 0, max: 1, palette: ['white', 'black']}, 'Missing Pixels');


// /* 3.3 Check band data types
// * Bands need to be the same data type in order to export 
// * multiband rasters to drive 
// */
 
print("Band Names", image_first.bandNames()); 
print("Band Types", image_first.bandTypes());

/* 4. Export Time Series to Google Drive
 * Export each image in the collection as multiband GeoTIFFs.
 */

/* Export parameters */
var folder = 'gee_exports';
var scale = 30; // 30-meter resolution
var crs = 'EPSG:4326'; // WGS 84 CRS

/* Define file naming function */
var fileNameFn = function(img) {
  var year = img.get('year').getInfo() || 'unknown';
  return 'landsat_multiband_' + year;
};

/* Export images to Google Drive */
utils.exportImageCollection(ls, aoi, folder, scale, crs, fileNameFn);


// /*
// * 6. Focal Analysis
// * Focal analyses and exporting original resolution (30 m) rasters. 
// */

/* 
* 6.1. Zero-meter focal (no smoothing) 
*/

// 6.1.1. No focal operation, keep original 30 m resolution
var ls_0 = ls.map(function(img) {
  var newNames = img.bandNames().map(function(name) {
    return ee.String(name).cat('_0');
  });
  return img.rename(newNames);
});

// print(ls_0, "ls_0");
// Map.addLayer(
//   ls_0.first().select('NDVI'), 
//   {}, 
//   'NDVI (0m focal, 30m res)'
// );

// 6.1.2. Define file naming function for 0 m exports
var fileNameFn_0 = function(img) {
  var year = img.get('year').getInfo() || 'unknown';
  return 'landsat_multiband_0_' + year;
};

// 6.1.3. Export each image to Google Drive at 30 m resolution
utils.exportImageCollection(
  ls_0,
  aoi,
  'gee_exports',
  990,          // scale in meters (30 m native resolution)
  'EPSG:3978', // CRS
  fileNameFn_0
);

/*
* 6.2. 150-meter focal analysis
*/
var kernelSize150 = 150;

// 6.2.1. Compute focal statistics at 30 m resolution
var ls_150 = ls.map(function(img) {
  return utils.focalStats(
    img, 
    kernelSize150, 
    'circle', 
    ['year']
  );
});

// print(ls_150, "ls_150");
// Map.addLayer(
//   ls_150.first().select('NDVI_mean_150'), 
//   {}, 
//   'NDVI (150m focal, 30m res)'
// );

// 6.2.2. Define the file naming function for 150 m exports
var fileNameFn_150 = function(img) {
  var year = img.get('year').getInfo() || 'unknown';
  return 'landsat_multiband_150_' + year;
};

// 6.2.3. Export the ls_150 collection at 30 m resolution
utils.exportImageCollection(
  ls_150,
  aoi,
  'gee_exports',
  990,          // scale in meters (30 m native resolution)
  'EPSG:3978', // CRS
  fileNameFn_150
);

/*
* 6.3. 250-meter focal analysis
*/
var kernelSize250 = 250;

// 6.3.1. Compute focal statistics at 30 m resolution
var ls_250 = ls.map(function(img) {
  return utils.focalStats(
    img, 
    kernelSize250, 
    'circle', 
    ['year']
  );
});

// print(ls_250, "ls_250");
// Map.addLayer(
//   ls_250.first().select('NDVI_mean_250'), 
//   {}, 
//   'NDVI (250m focal, 30m res)'
// );

// 6.3.2. Define the file naming function for 250 m exports
var fileNameFn_250 = function(img) {
  var year = img.get('year').getInfo() || 'unknown';
  return 'landsat_multiband_250_' + year;
};

// 6.3.3. Export the ls_250 collection at 30 m resolution
utils.exportImageCollection(
  ls_250,
  aoi,
  'gee_exports',
  990,          // scale in meters (30 m native resolution)
  'EPSG:3978', // CRS
  fileNameFn_250
);

File: scripts/preprocessing/gee/python/landsat_time_series.py
# ---
# title:   Landsat Time Series Analysis
# author:  Brendan Casey
# created: 2026-07-10
# inputs:
#   - Landsat 5/7/8/9 Surface Reflectance collections
#     (LANDSAT/*/C02/T1_L2)
#   - FAO GAUL province boundaries (Alberta)
# outputs:
#   - Annual multiband spectral-index GeoTIFFs exported to
#     Google Drive at native (30 m) resolution and at
#     focal scales (0/150/250 m) in EPSG:3978
#   - Per-band min/max summary CSV (image_stats)
# notes:
#   Python port of landsat_time_series.js for the Earth
#   Engine Python API. Builds an annual date list, computes
#   user-selected spectral indices via the shared ls_fn
#   helper, drops the QA_PIXEL band, casts to Float32, and
#   exports multiband images plus focal derivatives.
#
#   The original Map.addLayer/Map.centerObject calls, vis
#   parameters, and debug print() blocks are omitted.
#
#   Deviation: the shared ls_fn helper expects a client-
#   side list of date strings, so the ee.List produced by
#   create_date_list is materialized with getInfo() before
#   being passed in.
#
#   Setup (once):
#     pip install earthengine-api
#     earthengine authenticate
#   Then set EE_PROJECT in _gee_config.py and run.
# ---

import os
import sys

import ee

# Make utils importable regardless of the working
# directory VS Code runs the script from
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from _gee_config import DRIVE_FOLDER
from utils.compute_report import ComputeReport
from utils.gee_helpers import (
    calculate_image_collection_stats,
    create_date_list,
    export_image_collection,
    export_stats_to_csv,
    focal_stats,
)
from utils.gee_utils import initialize_ee
from utils.landsat_time_series import ls_fn

# 1. Setup ----

# 1.1 User parameters ----
LS_START_DATE = "2000-06-01"  # first time-series date
LS_END_DATE = "2024-06-01"  # last time-series date
LS_DATE_INTERVAL = 1  # step between series start dates
LS_DATE_INTERVAL_TYPE = "years"  # units for the step
LS_WINDOW = 121  # compositing window length
LS_WINDOW_TYPE = "days"  # units for the window
LS_STATISTIC = "mean"  # 'mean', 'median', 'max', etc.
LS_INDICES = [
    "BSI", "DRS", "DSWI", "EVI", "GNDVI",
    "LAI", "NBR", "NDMI", "NDSI", "NDVI",
    "NDWI", "SAVI", "SI",
]
EXPORT_SCALE = 30  # native Landsat resolution (m)
EXPORT_CRS = "EPSG:4326"  # native export CRS
FOCAL_SCALE = 990  # focal export scale (m)
FOCAL_CRS = "EPSG:3978"  # focal export CRS
FOCAL_KERNELS = [150, 250]  # focal radii (m), circle
PRINT_STATS = True  # min/max check (slow for large AOIs)
USE_TEST_AOI = True  # True: small test AOI; False: Alberta
COMPUTE_REPORT = True  # write EECU usage report (txt)

# 1.2 Initialize Earth Engine ----
# Project ID is read from _gee_config.py
initialize_ee()

# 1.3 Set up compute usage report ----
# Profiles EECU usage per section. Best used with
# USE_TEST_AOI = True to find choke points cheaply.
report = ComputeReport(
    "landsat_time_series",
    out_dir=os.path.join(
        os.path.dirname(os.path.abspath(__file__)),
        "gee_compute_reports",
    ),
    enabled=COMPUTE_REPORT,
)

# 2. Define study area ----
# Uses a small test polygon when USE_TEST_AOI is True;
# otherwise filters the FAO GAUL provinces for Alberta.

if USE_TEST_AOI:
    # Small aoi for testing purposes
    aoi = ee.Geometry.Polygon([
        [-113.5, 55.5],  # Top-left corner
        [-113.5, 55.0],  # Bottom-left corner
        [-112.8, 55.0],  # Bottom-right corner
        [-112.8, 55.5],  # Top-right corner
    ])
else:
    aoi = (
        ee.FeatureCollection(
            "FAO/GAUL_SIMPLIFIED_500m/2015/level1"
        )
        .filter(ee.Filter.eq("ADM0_NAME", "Canada"))
        .filter(ee.Filter.eq("ADM1_NAME", "Alberta"))
        .geometry()
    )

# 3. Build the time-series date list ----
# create_date_list returns an ee.List; ls_fn iterates a
# client-side list, so the dates are materialized as
# YYYY-MM-dd strings.
date_list = create_date_list(
    ee.Date(LS_START_DATE),
    ee.Date(LS_END_DATE),
    LS_DATE_INTERVAL,
    LS_DATE_INTERVAL_TYPE,
)
start_dates = (
    date_list.map(
        lambda d: ee.Date(d).format("YYYY-MM-dd")
    ).getInfo()
)

# 4. Landsat time-series processing ----
# Computes the selected spectral indices for each interval,
# drops the QA_PIXEL band, and casts every band to Float32.
ls = ls_fn(
    start_dates,
    LS_WINDOW,
    LS_WINDOW_TYPE,
    aoi,
    LS_INDICES,
    LS_STATISTIC,
)


def drop_qa_and_cast(image):
    """Drop the QA_PIXEL band and cast bands to Float32."""
    keep = image.bandNames().filter(
        ee.Filter.neq("item", "QA_PIXEL")
    )
    return image.select(keep).toFloat()


ls = ls.map(drop_qa_and_cast)

# 5. Check calculated bands (optional) ----
# Computes per-band min/max for the collection, prints the
# first-image summary, and exports the full table as a CSV.
if PRINT_STATS or COMPUTE_REPORT:
    reducer = ee.Reducer.min().combine(
        ee.Reducer.max(), "", True
    )
    collection_stats = calculate_image_collection_stats(
        ls, aoi, 1000, 1e13, reducer
    )
    with report.section("Landsat band min/max stats"):
        summary = (
            collection_stats.first()
            .toDictionary()
            .getInfo()
        )
    print("Landsat first-image stats:", summary)
    export_stats_to_csv(collection_stats, "image_stats")

# 6. Export time series to Google Drive ----
# Exports each image in the collection as a multiband
# GeoTIFF, one export task per image.


def landsat_file_name(img):
    """File name for the native-resolution export."""
    year = img.get("year").getInfo() or "unknown"
    return "landsat_multiband_" + str(year)


export_image_collection(
    ls,
    aoi,
    DRIVE_FOLDER,
    EXPORT_SCALE,
    EXPORT_CRS,
    landsat_file_name,
)

# 7. Focal analysis ----
# Exports focal (neighbourhood) statistics at 0/150/250 m
# in EPSG:3978. The 0 m case renames bands with a "_0"
# suffix but applies no smoothing.

# 7.1 Zero-metre focal (no smoothing) ----


def rename_zero_focal(img):
    """Append a "_0" suffix to every band name."""
    new_names = img.bandNames().map(
        lambda name: ee.String(name).cat("_0")
    )
    return img.rename(new_names)


ls_0 = ls.map(rename_zero_focal)


def landsat_file_name_0(img):
    """File name for the 0 m focal export."""
    year = img.get("year").getInfo() or "unknown"
    return "landsat_multiband_0_" + str(year)


export_image_collection(
    ls_0,
    aoi,
    DRIVE_FOLDER,
    FOCAL_SCALE,
    FOCAL_CRS,
    landsat_file_name_0,
)

# 7.2 Circular focal means (150 m, 250 m) ----
for kernel_size in FOCAL_KERNELS:
    ls_focal = ls.map(
        lambda img, k=kernel_size: focal_stats(
            img, k, "circle", ["year"]
        )
    )

    def make_focal_file_name(k):
        def focal_file_name(img):
            year = img.get("year").getInfo() or "unknown"
            return "landsat_multiband_" + str(k) + "_" + str(
                year
            )

        return focal_file_name

    export_image_collection(
        ls_focal,
        aoi,
        DRIVE_FOLDER,
        FOCAL_SCALE,
        FOCAL_CRS,
        make_focal_file_name(kernel_size),
    )

# 8. Compute usage report ----
# Writes the profiled sections to gee_compute_reports/.
# Collection exports start many batch tasks, so per-task
# EECU totals are not logged here; monitor progress at
# https://code.earthengine.google.com/tasks
report.write()

# End of script ----
